src/Essentials/src/FileSystem/FileSystem.shared.cs

```diff
@@ -1,26 +1,67 @@
 using System;
 using System.IO;
 using System.Threading.Tasks;
+using System.ComponentModel;
+using Microsoft.Maui.Essentials;
+using Microsoft.Maui.Essentials.Implementations;
 
 namespace Microsoft.Maui.Essentials
 {
+	public interface IFileSystem
+	{
+		string CacheDirectory { get; }
+
+		string AppDataDirectory { get; }
+
+		Task<Stream> OpenAppPackageFileAsync(string filename);
+
+		Task<bool> AppPackageFileExistsAsync(string filename);
+	}
+
+	public interface IPlatformFileSystem
+	{
+#if ANDROID
+		Java.IO.File GetTemporaryFile(Java.IO.File root, string fileName);
+
+		string EnsurePhysicalPath(Android.Net.Uri uri, bool requireExtendedAccess = true);
+#endif
+#if IOS || MACCATALYST
+		Task<FileResult[]> EnsurePhysicalFileResultsAsync(params Foundation.NSUrl[] urls);
+#endif
+	}
+
 	/// <include file="../../docs/Microsoft.Maui.Essentials/FileSystem.xml" path="Type[@FullName='Microsoft.Maui.Essentials.FileSystem']/Docs" />
-	public static partial class FileSystem
+	public static class FileSystem
 	{
 		/// <include file="../../docs/Microsoft.Maui.Essentials/FileSystem.xml" path="//Member[@MemberName='CacheDirectory']/Docs" />
 		public static string CacheDirectory
-			=> PlatformCacheDirectory;
+			=> Current.CacheDirectory;
 
 		/// <include file="../../docs/Microsoft.Maui.Essentials/FileSystem.xml" path="//Member[@MemberName='AppDataDirectory']/Docs" />
 		public static string AppDataDirectory
-			=> PlatformAppDataDirectory;
+			=> Current.AppDataDirectory;
 
 		/// <include file="../../docs/Microsoft.Maui.Essentials/FileSystem.xml" path="//Member[@MemberName='OpenAppPackageFileAsync']/Docs" />
 		public static Task<Stream> OpenAppPackageFileAsync(string filename)
-			=> PlatformOpenAppPackageFileAsync(filename);
+			=> Current.OpenAppPackageFileAsync(filename);
 
 		public static Task<bool> AppPackageFileExistsAsync(string filename)
-			=> PlatformAppPackageFileExistsAsync(filename);
+			=> Current.AppPackageFileExistsAsync(filename);
+
+
+#if ANDROID
+		internal static  Java.IO.File GetTemporaryFile(Java.IO.File root, string fileName)
+			=> (Current as IPlatformFileSystem)?.GetTemporaryFile(root, fileName);
+
+		internal static string EnsurePhysicalPath(Android.Net.Uri uri, bool requireExtendedAccess = true)
+			=> (Current as IPlatformFileSystem)?.EnsurePhysicalPath(uri, requireExtendedAccess);
+#endif
+
+#if IOS || MACCATALYST
+		internal static Task<FileResult[]> EnsurePhysicalFileResultsAsync(params Foundation.NSUrl[] urls)
+			=> (Current as IPlatformFileSystem)?.EnsurePhysicalFileResultsAsync(urls);
+#endif
```

Comment:
I wonder if these should throw? Especially the async... a null Task is very unexpected.